packages/network/src/logic/receipts/ReceiptRequester.ts

```diff
@@ -0,0 +1,145 @@
+    receiptStore: ReceiptStore
+    signers: Signers
+    windowTimeoutMargin?: number
+    bucketUpdateTimeoutMargin?: number
+}
+
+/**
+ * Plays the active role of the receipt requester.
+ *
+ * When a bucket is closed (due to window timeout or bucket update timeout),
+ * this class will send a receipt request to the counterparty node in order to
+ * establish a common ground truth on the amount of stream message data this
+ * node has sent and what the counterparty has received during the window.
+ *
+ * It will then wait for a response, and upon receiving such, will perform
+ * signature validations and such. If everything looks good, it will ask
+ * `ReceiptStore` to store the receipt.
+ */
+export class ReceiptRequester {
+    private readonly myNodeId: NodeId
+    private readonly nodeToNode: Pick<NodeToNode, 'on' | 'send' | 'registerErrorHandler'>
+    private readonly receiptStore: ReceiptStore
+    private readonly signers: Signers
+    private readonly windowTimeoutMargin: number
+    private readonly bucketUpdateTimeoutMargin: number
+    private readonly debouncedTasks = new DebouncedTaskManager<BucketID>()
+    private readonly collector = new BucketCollector((bucket) => {
+        this.debouncedTasks.schedule(bucket.getId(), this.getCloseTime(bucket) - Date.now(), () => {
+            this.sendReceiptRequest(bucket)
+            this.collector.removeBucket(bucket.getId())
+            logger.info('closed bucket %s', bucket.getId())
+        })
+    })
+
+    constructor({
+        myNodeId,
+        nodeToNode,
+        receiptStore,
+        signers,
+        windowTimeoutMargin,
+        bucketUpdateTimeoutMargin
+    }: ConstructorOptions) {
+        this.myNodeId = myNodeId
+        this.nodeToNode = nodeToNode
+        this.receiptStore = receiptStore
+        this.signers = signers
+        this.windowTimeoutMargin = windowTimeoutMargin || DEFAULT_WINDOW_TIMEOUT_MARGIN
+        this.bucketUpdateTimeoutMargin = bucketUpdateTimeoutMargin || DEFAULT_UPDATE_TIMEOUT_MARGIN
+
+        nodeToNode.on(Event.BROADCAST_MESSAGE_SENT, ({ streamMessage }, recipient) => {
+            this.collector.record(streamMessage, recipient)
+        })
+        nodeToNode.on(Event.RECEIPT_RESPONSE_RECEIVED, ({ receipt }, source) => {
+            const { receiver, sender } = receipt.claim
+            if (receiver !== source) { // TODO: This is not necessarily wrong, however, why would these get relayed thru a 3rd party?
+                logger.warn('unexpected receipt from %s (trying to respond on behalf of %s)', source, receiver)
+                // TODO: cut connection?
+                return
+            }
+            if (sender !== this.myNodeId) {
+                logger.warn('unexpected receipt from %s (the claim directed to %s does not concern me)',
+                    source, sender)
+                // TODO: cut connection?
+                return
+            }
+            if (!this.signers.receipt.validate(receipt)) {
```

Comment:
Does this also validate the claim itself? I.e. is the if-check after this one necessary?